CONTRIBUTING.md

```diff
@@ -1,73 +1,76 @@
+
+# C++
+* Use automatic memory management (RAII, smart pointers). `new` and `malloc` are evil words.
+* Use `const` whenever possible. Consider variables, methods, pointers, and their pointees.
+* Reduce the size of your hpp files, both in terms of the number of lines and the code complexity. This keeps the compilation duration low.
+  * Code in hpp files is compiled for every cpp file that includes the header. As such, move code to cpp files where possible. This often includes templated classes, where it is sometimes possible to implement their code in cpp files.
+  * Anonymous namespaces are a good way to define local helper methods.
+  * Use forward declarations instead of full header includes wherever possible.
+* Loops
+  * Use range-based for loops when possible: `for (const auto& item : items) {...}`.
+  * If you have to use old-style loops, keep in mind that the loop condition is evaluated every time: Instead of `for (auto i = 0; i < expensive_function(); ++i)`, the comparison value should be calculated just once. http://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop
+* Data structures
+  * When creating a vector where you know the size beforehand, use `reserve` to avoid unnecessary resizes and allocations.
+  * Hash-based data structures are usually faster than tree-based data structures. Unless you have a reason to use the latter, prefer unordered_(map|set) over map and set.
+* Copies
+
+    * C++ makes it too easy to inadvertently copy objects.
+
+    * For larger elements (e.g., vectors), pass a (const) reference instead.
+
+    * When your implemented class does not need to be copied (e.g., a Table should never exist more than once), inherit from Noncopyable to avoid these potentially expensive copies.
+
+* Miscellaneous
+  * Prefer `if (object) {` over `if (object != nullptr) {` or `if (object.has_value()) {`.
+  * Don't write `this->` if you don't have to.
+  * Be explicit with types: Use [u]int(8|16|32|64)_t instead of `int, long, uint` etc.
+  * Use auto-to-stick (https://www.fluentcpp.com/2018/09/28/auto-stick-changing-style/): `auto x = 17;` or `auto y = std::vector<size_t>{};`.
+  * Namespaces: Do not create nested namespaces, do not import namespaces.
+
+# Formatting and Naming
+* Much of this is enforced by clang-tidy. However, clang-tidy does not yet cover hpp files (see #1901). Also, while clang-tidy is a great help, do not rely on it.
+* Call ./scripts/format.sh before pushing.
+* Choose clear and concise names, and avoid, e.g., `i`, `j`, `ch_ptr`.
+* Formatting details: 2 spaces, 120 columns, comments above code.
+* Use empty lines to structure your code.
+* Naming conventions:
+
+    * Files: lowercase separated by underscores, e.g., abstract_operator.cpp, usually corresponding to a class, e.g., AbstractOperator.
+
+    * Types (classes, structs, enums, typedefs, using): CamelCase starting with uppercase letter, e.g., `TableScan`.
```

Comment:
I thought this naming style was referred to as PascalCase, not camelCase (https://en.wikipedia.org/wiki/Camel_case).

Reply:
I'm fine with both.

Reply:
Changed to PascalCase.